src/layers/attention.rs

```diff
@@ -52,16 +52,143 @@ pub enum QkvMode {
+        Ok(Self {
+            slopes,
+            is_causal,
+            is_inverted,
+        })
+    }
+
+    /// Calculate the linear bias slopes for a given number
+    /// of attention heads.
+    ///
+    /// * n_attention_heads - Number of attention heads.
+    ///
+    /// Returns: Head slope tensor.
+    /// *Shape:* `(1, heads, 1, 1)`
+    #[allow(clippy::unnecessary_cast)]
+    fn calculate_slopes(n_attention_heads: usize) -> Result<Tensor> {
+        fn slopes_with_step(n_attention_heads: usize, step: usize) -> Result<Tensor> {
+            let ratio = 2.0f32.powf(-8.0f32 / n_attention_heads as f32);
+            let slope = (1..(1 + n_attention_heads))
+                .step_by(step)
+                .map(|x| ratio.powi(x as i32))
+                .collect::<Vec<_>>();
+            Ok(Tensor::new(slope, &Device::Cpu)?)
+        }
+
+        // The slope as proposed in the ALiBi paper would be:
+        //
+        // return slopes_with_step(n_attention_heads)
+        //
+        // However the authors note in their implementation that using powers
+        // of 2 for n in the ratio 2**(-8/n) of the geometric sequence for
+        // slopes has better properties.
+        //
+        // Most implementations use powers of two in the following
+        // manner: if the number of heads is not a power of 2, then we find
+        // k=the largest power of 2 in 1..n. The slopes are then computed
+        // as the concatenation of:
+        //
+        // - The slopes for 1..k.
+        // - The slopes for 1..2*k with step 2, taking the first n-k elements.
+
+        // k is the largest power of 2 in 1..n.
+        // Coerced to the same type here to make the bit arithmetic explicit.
+        let k = 1 << ((usize::BITS - (n_attention_heads as usize).leading_zeros()) - 1);
+        let mut slopes = slopes_with_step(k, 1)?;
+
+        if n_attention_heads != k {
+            let remaining = n_attention_heads - k;
+            let slopes_rest = slopes_with_step(2 * k, 2)?.i(..remaining)?;
+            slopes = Tensor::cat(&[slopes, slopes_rest], 0)?;
+        }
+
+        Ok(slopes.reshape((1, (), 1, 1))?)
+    }
+
+    /// Calculate the linear bias tensor upto a given (key) sequence length.
+    ///
+    /// * seq_len - Maximum number of timesteps to calculate.
+    ///
+    /// Returns: Multi-headed linear bias tensor.
+    /// *Shape:* `(1, heads, seq_len, seq_len)` (non-causal) or
+    /// `(1, heads, 1, seq_len)` (causal)
+    fn calculate_biases(&self, seq_len: usize) -> Result<Tensor> {
+        let mut distances = match self.is_causal {
+            true => Tensor::arange(1 - seq_len as i64, 1, &Device::Cpu)?,
+            false => {
+                let mut distance = Tensor::arange(0, seq_len as i64, &Device::Cpu)?;
+                distance = distance.broadcast_sub(&distance.reshape(((), 1))?)?;
+                distance
+                    .abs()?
+                    .broadcast_mul(&Tensor::new(-1i64, &Device::Cpu)?)?
```

Comment:
As in no vectorized implementation? Does it get coerced to another type internally (the tests appear to be working)?

Reply:
I was more lamenting that there is no implementation for `Tensor` of `Mul` taking i64 (only floating point types). Would be nice if candle had that, because then we could just say `distance.abs()? * -1i64`.

Reply:
Ah, right. Yeah, having to manually allocate `Tensor`s for non-fp scalars is indeed meh.